crates/linker/src/pe.rs

```diff
@@ -1,47 +1,343 @@
-use std::{ops::Range, path::Path, time::Instant};
+use std::{
+    io::{BufReader, BufWriter},
+    ops::Range,
+    path::Path,
+    time::Instant,
+};
 
-use memmap2::{Mmap, MmapMut};
+use bincode::{deserialize_from, serialize_into};
+use memmap2::MmapMut;
 use object::{
-    pe::{ImageImportDescriptor, ImageNtHeaders64, ImageSectionHeader, ImageThunkData64},
+    pe::{
+        self, ImageFileHeader, ImageImportDescriptor, ImageNtHeaders64, ImageSectionHeader,
+        ImageThunkData64,
+    },
     read::pe::ImportTable,
-    LittleEndian as LE, RelocationTarget, SectionIndex,
+    LittleEndian as LE, Object, RelocationTarget, SectionIndex,
 };
+use serde::{Deserialize, Serialize};
+
 use roc_collections::MutMap;
 use roc_error_macros::internal_error;
 
-use crate::{load_struct_inplace, load_struct_inplace_mut};
+use crate::{generate_dylib::APP_DLL, load_struct_inplace, load_struct_inplace_mut};
+
+#[derive(Debug, Serialize, Deserialize)]
+struct PeMetadata {
```

Comment:
I'm trying to piece together the context here. What I think I'm seeing is:
We are making a binary file in our own custom format, defined by this data structure, and writing it out to disk to cache the results of some up-front analysis work we are doing on the host. This gives the app developer a faster development cycle.
So the precompiled host effectively exists in two files. The executable and the metadata. And some part of our compiler is checking that the metadata has a later timestamp than the executable.

Is that right?
This might be a good place for a doc comment explaining that kind of context.

Reply:
yes, we have two files

- precompiledhost: the host executable with space for additional sections 
- metadata: various offsets into the precompiledhost

We want our linking to be as fast as possible, which means that we want to shift as much work as possible to the preprocessing phase.
That in turn means we erase some information from the host executable that we need later (e.g. about our dummy dll). 

We improve performance further by collecting a bunch of information about the binary in the metadata. That means we don't have to parse the precompiledhost when doing the final linking.

Then, right now we always rebuild the host. You need to pass the `--precompiled-platform` stop it from rebuilding. The only caching we do is of the dummy .dll, where if the on-disk .dll already has exactly the symbols that we need, we don't write it again. 

I'll find some place for this information

Reply:
Nice, great comment